app/src/main/java/org/jellyfin/androidtv/data/eventhandling/SocketHandler.kt

```diff
@@ -0,0 +1,164 @@
+		addListener<LibraryChangedMessage> { message -> onLibraryChanged(message.info) }
+
+		// Media playback
+		addListener<PlayMessage> { message -> onPlayMessage(message) }
+		addListener<PlayStateMessage> { message -> onPlayStateMessage(message) }
+
+		// General commands
+		addGeneralCommandsListener(setOf(GeneralCommandType.DISPLAY_CONTENT)) { message ->
+			val itemId by message
+			val itemUuid = itemId?.toUUIDOrNull()
+
+			if (itemUuid != null) onDisplayContent(itemUuid)
+		}
+		addGeneralCommandsListener(setOf(GeneralCommandType.DISPLAY_MESSAGE, GeneralCommandType.SEND_STRING)) { message ->
+			val header by message
+			val text by message
+			val string by message
+
+			onDisplayMessage(header, text ?: string)
+		}
+	}
+
+	private fun onLibraryChanged(info: LibraryUpdateInfo) {
+		Timber.d(buildString {
+			appendLine("Library changed.")
+			appendLine("Added ${info.itemsAdded!!.size} items")
+			appendLine("Removed ${info.itemsRemoved!!.size} items")
+			appendLine("Updated ${info.itemsUpdated!!.size} items")
+		})
+
+		if (info.itemsAdded!!.any() || info.itemsRemoved!!.any())
+			dataRefreshService.lastLibraryChange = System.currentTimeMillis()
+	}
+
+	private fun onPlayMessage(message: PlayMessage) {
+		val itemId = message.request.itemIds?.firstOrNull() ?: return
+
+		PlaybackHelper.retrieveAndPlay(
+			itemId.toString(),
+			false,
+			message.request.startPositionTicks,
+			context
+		)
+	}
+
+	@Suppress("ComplexMethod")
+	private fun onPlayStateMessage(message: PlayStateMessage) = coroutineScope.launch(Dispatchers.Main) {
+		Timber.i("Received PlayStateMessage with command ${message.request.command}")
+		val playbackController = TvApp.getApplication()?.playbackController
+		// Audio playback uses the mediaManager, video playback and live tv use the playbackController
+		if (mediaManager.isAudioPlayerInitialized) when (message.request.command) {
+			PlaystateCommand.STOP -> mediaManager.stopAudio(true)
+			PlaystateCommand.PAUSE, PlaystateCommand.UNPAUSE, PlaystateCommand.PLAY_PAUSE -> mediaManager.playPauseAudio()
+			PlaystateCommand.NEXT_TRACK -> mediaManager.nextAudioItem()
+			PlaystateCommand.PREVIOUS_TRACK -> mediaManager.prevAudioItem()
+			// Not implemented
+			PlaystateCommand.SEEK,
+			PlaystateCommand.REWIND,
+			PlaystateCommand.FAST_FORWARD -> Unit
+		} else when (message.request.command) {
+			PlaystateCommand.STOP -> playbackController?.endPlayback(true)
+			PlaystateCommand.PAUSE, PlaystateCommand.UNPAUSE, PlaystateCommand.PLAY_PAUSE -> playbackController?.playPause()
+			PlaystateCommand.NEXT_TRACK -> playbackController?.next()
+			PlaystateCommand.PREVIOUS_TRACK -> playbackController?.prev()
+			PlaystateCommand.SEEK -> playbackController?.seek(
+				(message.request.seekPositionTicks ?: 0) / TICKS_TO_MS
+			)
+			// FIXME get rewind/forward amount from displayprefs
+			PlaystateCommand.REWIND -> playbackController?.skip(REWIND_MS)
+			PlaystateCommand.FAST_FORWARD -> playbackController?.skip(FORWARD_MS)
```

Comment:
So this is a bit of a code smell to me. The socket handler should have no idea how to handle a playback command (IMO). Instead it should forward to the relevant controller and let that perform the dispatch.

Another good reason is future developers who want to implement `REWIND` for audio won't necessarily look in `SocketHandler` for the dispatch, personally I'd look at the audio controller first.

Reply:
There is a reason for the "FIXME" comment. I don't work on the old playback code, all SDK migration work is necessary for the playback rewrite (#1057).
The new playback module will add it's own listeners to the SDK to handle playstate commands. This PR just migrates the existing stuff and doesn't focus on fixing the current playback spaghetti code.

Reply:
Ah I couldn't see the FIXME comment, probably a good reason not to skim/review stuff at 3AM...

:+1: Yeah I'm all for not touching the spaghetti code where we can avoid it